### PR TITLE
Bugfix: Use one type of Raft logs and remove distributed locks.

### PR DESCRIPTION
**Pull request by as51340**
Coordinators now use a single type of raft logs (update_cluster_state) to remove the need for distributed locks. Coordinators can now be in 3 states: follower, leader not ready to react to instance callbacks and leader which is ready to react to instance callbacks. When a coordinator becomes the leader, it sets its state to leader not ready and then changes it to ready only after reconciliation finishes.

### CODE REVIEW

**Comment by as51340** [comment]
### Tracking
- [ ] **[Link to Epic/Issue]**


### Standard development
- [ ] Update unit/E2E tests
- [ ] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch this branch


### CI Testing Labels
- [ ] Select the appropriate CI test labels _(CI -build=**build-name** -test=**test-suite**)_


### Documentation checklist
- [ ] Add the documentation label
- [ ] Add the bug / feature label
- [ ] Add the milestone for which this feature is intended
    - If not known, set for a later milestone
- [ ] Write a release note, including added/changed clauses
    - **[ Release note text ]**
- [ ] **[ Documentation PR link memgraph/documentation#XXXX ]**
    - [ ] Is back linked to this development PR
- [ ] **[ Tag someone from docs team ]**

File: src/coordination/coordinator_instance.cpp [hunk 1]
```diff
@@ -426,25 +418,24 @@ auto CoordinatorInstance::ReconcileClusterState_() -> ReconcileClusterStateStatu
...
+  // Shouldn't execute
   return ReconcileClusterStateStatus::SUCCESS;
 }
 
 auto CoordinatorInstance::TryVerifyOrCorrectClusterState() -> ReconcileClusterStateStatus {
   // Follows nomenclature from replication handler where Try<> means doing action from the
   // user query.
+  auto expected = CoordinatorStatus::LEADER_READY;
+  if (!status.compare_exchange_strong(expected, CoordinatorStatus::LEADER_NOT_READY)) {
+    return ReconcileClusterStateStatus::NOT_LEADER_ANYMORE;
```
[the remark this answers is not in the document]
Comment (as51340) [reply src/coordination/coordinator_instance.cpp:430]: in that case I don't want to allow user to reconcile state

File: src/coordination/coordinator_instance.cpp [hunk 2]
```diff
@@ -301,23 +302,29 @@ auto CoordinatorInstance::ShowInstances() const -> std::vector<InstanceStatus> {
...
-        spdlog::trace("Reconcile cluster state finished successfully.");
+      case (ReconcileClusterStateStatus::SUCCESS): {
+        auto expected = CoordinatorStatus::LEADER_NOT_READY;
+        if (!status.compare_exchange_strong(expected, CoordinatorStatus::LEADER_READY)) {
+          if (expected == CoordinatorStatus::FOLLOWER) {
+            spdlog::trace("Reconcile cluster state finished successfully but coordinator isn't leader anymore.");
+            return ReconcileClusterStateStatus::NOT_LEADER_ANYMORE;
+          }
+          // We should never get into such state but we log it for observability.
+          spdlog::trace("Reconcile cluster state finished successfully but coordinator is already in state ready.");
```
[the remark this answers is not in the document]
Comment (as51340) [reply src/coordination/coordinator_instance.cpp:319]: true but I think without consequences?

**Comment by hal-eisen-MG** [comment]
@as51340 Where can I learn more about "Coordinator log store needs to be started from scratch"? I'd like to explain to the user how to do that. Maybe an anchor on this page? https://memgraph.com/docs/clustering/high-availability

**Comment by as51340** [comment]
> @as51340 Where can I learn more about "Coordinator log store needs to be started from scratch"? I'd like to explain to the user how to do that. Maybe an anchor on this page? https://memgraph.com/docs/clustering/high-availability

It means delete data_directory/network and data_directory/logs and restart memgraph

**Review by andrejtonev — APPROVED** [review]
Great work!
Approved even with the known limitations as the problems have been documented as issues on gh.

**Comment by andrejtonev** [comment]
@as51340 you need to solve the sonarcloud issues before you can merge.

**Pull request merged**